src/shared/bpf-dlopen.c

```diff
@@ -67,8 +72,32 @@ int dlopen_bpf(void) {
                         DLSYM_ARG(libbpf_probe_bpf_prog_type),
                         DLSYM_ARG(libbpf_set_print),
                         DLSYM_ARG(libbpf_get_error));
-        if (r < 0)
-                return r;
+        if (r < 0) {
+                r = dlopen_many_sym_or_warn(
+                                &bpf_dl, "libbpf.so.0", LOG_DEBUG,
+                                DLSYM_ARG(bpf_link__destroy),
+                                DLSYM_ARG(bpf_link__fd),
+                                DLSYM_ARG(bpf_map__fd),
+                                DLSYM_ARG(bpf_map__name),
+                                DLSYM_ARG(bpf_create_map),
+                                DLSYM_ARG(bpf_map__set_max_entries),
+                                DLSYM_ARG(bpf_map_update_elem),
+                                DLSYM_ARG(bpf_map_delete_elem),
+                                DLSYM_ARG(bpf_map__set_inner_map_fd),
+                                DLSYM_ARG(bpf_object__open_skeleton),
+                                DLSYM_ARG(bpf_object__load_skeleton),
+                                DLSYM_ARG(bpf_object__attach_skeleton),
+                                DLSYM_ARG(bpf_object__detach_skeleton),
+                                DLSYM_ARG(bpf_object__destroy_skeleton),
+                                DLSYM_ARG(bpf_probe_prog_type),
+                                DLSYM_ARG(bpf_program__attach_cgroup),
+                                DLSYM_ARG(bpf_program__attach_lsm),
+                                DLSYM_ARG(bpf_program__name),
+                                DLSYM_ARG(libbpf_set_print),
+                                DLSYM_ARG(libbpf_get_error));
```

Comment:
please do not duplicate the whole shebang here. 

Instead, use dlopen() to open one or the other .so, and then use dlsym_many_or_warn() instead of dlopen_many_sym_or_warn() to resolve the symbols.

(I presume the list of symbols is exactly the same?)

Reply:
Unfortunately, no; the list is not exactly the same.
`bpf_map_create` and `libbpf_probe_bpf_prog_type` got added in 0.6 and 0.7 respectively (released in Dec 2021 and Feb 2022), and many distros haven't updated yet to these so the direction was towards compat stuff.

If 0.7.0 is acceptable, then dlopen and dlsym_many_or_warn would be much easier, and all the compat cruft can be removed

Reply:
well, we still can have a single dlsym_many_or_warn() for the common parts, and then for the disjunct part have another one.